contracts/Staking.sol

```diff
@@ -0,0 +1,507 @@
+        _stakeFor(_account, _amount, _data);
+    }
+
+    /**
+     * @notice Unstakes `_amount` tokens, returning them to the user
+     * @param _amount Number of tokens staked
+     * @param _data Used in Unstaked event, to add signalling information in more complex staking applications
+     */
+    function unstake(uint256 _amount, bytes _data) external {
+        // unstaking 0 tokens is not allowed
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // transfer tokens
+        // it will fail for non ERC20 compliant tokens which don't return anything
+        // see: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
+        require(stakingToken.transfer(msg.sender, _amount), ERROR_TOKEN_TRANSFER);
+
+        // checkpoint updated staking balance
+        _modifyStakeBalance(msg.sender, _amount, false);
+
+        // checkpoint total supply
+        _modifyTotalStaked(_amount, false);
+
+        emit Unstaked(msg.sender, _amount, totalStakedFor(msg.sender), _data);
+    }
+
+    /**
+     * @notice Lock `_amount` staked tokens and assign `_manager` as manager with `_data` as data, so they can not be unstaked
+     * @param _amount The amount of tokens to be locked
+     * @param _manager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
+     * @param _data Data to parametrize logic for the lock to be enforced by the manager
+     * @return The id of the newly created lock
+     */
+    function lock(uint256 _amount, address _manager, bytes _data) external returns (uint256) {
+        // locking 0 tokens is invalid
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // check enough unlocked tokens are available
+        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
+
+        accounts[msg.sender].lastLockId++;
+        uint256 _lockId = accounts[msg.sender].lastLockId;
+        accounts[msg.sender].activeLockIds.push(_lockId);
+        Lock storage lock_ = accounts[msg.sender].locks[_lockId];
+        lock_.amount = _amount;
+        lock_.unlockedAt = MAX_UINT64;
+        lock_.manager = ILockManager(_manager);
+        lock_.data = _data;
+
+        emit Locked(msg.sender, _lockId, _amount, _manager, _data);
+
+        return _lockId;
+    }
+
+    /**
+     * @notice Try to unlock as much locks belonging to `_account` as possible
+     * @dev It won't work (it will revert) if one of the managers is an EOA
+     * @param _account Owner whose locks are to be unlocked
+     */
+    function unlockAll(address _account) external {
+        Account storage account = accounts[_account];
+
+        for (uint256 i = account.activeLockIds.length; i > 0; i--) {
+            if (canUnlock(_account, account.activeLockIds[i - 1])) {
+                unlock(_account, account.activeLockIds[i - 1]);
+            }
+        }
```

Comment:
Yes, the reason is that the other way around doesn't work, because on `unlock` we are removing elements from the array and therefore reducing the length of the array (which would be checked on the condition part of the for loop).
About this:
> we are performing two unnecessary subtractions per iteration
It was due to doing it backwards and with `uint`, but I guess I could use a signed int for the iteration.